fdbclient/MultiVersionTransaction.actor.cpp

```diff
@@ -2052,98 +2093,100 @@ ThreadFuture<Void> MultiVersionDatabase::DatabaseState::monitorProtocolVersion()
+	if (itr == clients.end()) {
+		// We don't have a client matching the current protocol
+		initializationState = InitializationState::INCOMPATIBLE;
+		updateDatabase(Reference<IDatabase>(), Reference<ClientInfo>());
+		return;
+	}
 
-				// Put the client in a disconnected state until the version changes again
-				updateDatabase(Reference<IDatabase>(), Reference<ClientInfo>());
-				return;
-			}
+	// A compatible client found, use it for creating a new database connection
+	auto& client = itr->second;
+	TraceEvent("CreatingDatabaseOnClient")
+	    .detail("LibraryPath", client->libPath)
+	    .detail("Failed", client->failed)
+	    .detail("External", client->external);
 
-			if (client->external && !MultiVersionApi::api->getApiVersion().hasInlineUpdateDatabase()) {
-				// Old API versions return a future when creating the database, so we need to wait for it
-				Reference<DatabaseState> self = Reference<DatabaseState>::addRef(this);
-				dbReady = mapThreadFuture<Void, Void>(
-				    newDb.castTo<DLDatabase>()->onReady(), [self, newDb, client](ErrorOr<Void> ready) {
-					    if (!ready.isError()) {
-						    onMainThreadVoid([self, newDb, client]() { self->updateDatabase(newDb, client); });
-					    } else {
-						    onMainThreadVoid(
-						        [self, client]() { self->updateDatabase(Reference<IDatabase>(), client); });
-					    }
+	Reference<IDatabase> newDb;
+	try {
+		newDb = connectionRecord.createDatabase(client->api);
+	} catch (Error& e) {
+		// Create error currently does not return any error except for network not initialized,
+		// which cannot happen at this point
+		ASSERT(false);
+	}
 
-					    return ready;
-				    });
-			} else {
-				updateDatabase(newDb, client);
-			}
-		} else {
-			// We don't have a client matching the current protocol
-			updateDatabase(Reference<IDatabase>(), Reference<ClientInfo>());
-		}
+	if (client->external && !MultiVersionApi::api->getApiVersion().hasInlineUpdateDatabase()) {
+		// Old API versions return a future when creating the database, so we need to wait for it
+		Reference<DatabaseState> self = Reference<DatabaseState>::addRef(this);
+		dbReady = mapThreadFuture<Void, Void>(
+		    newDb.castTo<DLDatabase>()->onReady(), [self, newDb, client](ErrorOr<Void> ready) {
+			    if (!ready.isError()) {
+				    onMainThreadVoid([self, newDb, client]() { self->updateDatabase(newDb, client); });
+			    } else {
+				    onMainThreadVoid(
+				        [self, client]() { self->updateDatabase(Reference<IDatabase>(), Reference<ClientInfo>()); });
+			    }
+			    return ready;
+		    });
+	} else {
+		updateDatabase(newDb, client);
 	}
 }
 
 // Replaces the active database connection with a new one. Must be called from the main thread.
 void MultiVersionDatabase::DatabaseState::updateDatabase(Reference<IDatabase> newDb, Reference<ClientInfo> client) {
-	if (closed) {
+	if (initializationState == InitializationState::CLOSED) {
```

Comment:
Can we remove `closed` from the header file too now?

Reply:
Oh I yes, I thought I did